src/mlpack/methods/ann/loss_functions/poisson_nll_loss_impl.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file poisson_nll_loss_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Poisson NegativeLogLikelihood class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+
+
+// In case it hasn't yet been included.
+#include "poisson_nll_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PoissonNLLLoss<InputDataType, OutputDataType>::PoissonNLLLoss(
+  const bool log_input,
+  const bool full,
+  const double eps,
+  const bool reduce):
+  log_input(log_input), full(full), eps(eps), reduce(reduce)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double PoissonNLLLoss<InputDataType, OutputDataType>::Forward(
+  const InputType&& input,
+  TargetType&& target)
+{
+  InputType loss(size(input));
+  if (log_input)
+  {
+    loss = arma::exp(input) - target % input;
+  }
+  else
+  {
+    loss = input - target % arma::log(input + eps);
```

Comment:
Hey @mrityunjay-tripathi , I think `input` is of type InputTypes and eps is of type double. Is it possible to do 
```
inputs + eps;
```
It should be like this 
```
loss = input - target % arma::log(size(input) + eps);
```
Let me know if I am missing something

Reply:
Hi @himanshupathak21061998 , @mrityunjay-tripathi, I think expression written above should have the same meaning and 
`loss = input - target % arma::log(size(input) + eps);`
I think there is something wrong with implementation of the current code as size(input) has no physical meaning.
You can refer [this pytorch implementation](https://github.com/pytorch/pytorch/pull/19316/files#diff-454c8378b774ddef9050815381b85da9R146).

Reply:
Yes @kartikdutt18 I was wrong thanks for clarification.

Reply:
Not an issue, I think you could declare an an` InputType matEps(input.size(), eps) `and then use it.

Reply:
I think this line is fine and serves the purpose, i.e. ```loss = input - target * log(input + eps)```. And yes I think we can perform the addition of ```input``` and ```eps```. The demo code can be -
```
#include<bits/stdc++.h>
#include<armadillo>

using namespace std;

int main()
{
	arma::mat input;
	input = arma::mat("1.0 1.0 3.0 2.0 1.0 1.0 2.0 0.0");
	input.print();
	input = arma::log(input + 0.1);
	input.print();
}
```